### DISCUSSION TITLE
Form::isValid() returning true despite invalid EmailType value

### DISCUSSION THREAD

**Discussion by php4fan** [symfony/symfony#44055] [Q&A] [answered]
I have a **form type** like this:

```
<?php

namespace App\Form;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\EmailType;
use Symfony\Component\Form\Extension\Core\Type\PasswordType;
use Symfony\Component\Form\FormBuilderInterface;

class RegisterFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('email', EmailType::class)
            ->add('password', PasswordType::class)
            ->add('name')
        ;
    }
}
```

Then in my **controller**:
```
<?php

namespace App\Controller;

use App\Form\RegisterFormType;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Request;
//...

class UserController extends AbstractController
{
    public function register(EntityManagerInterface $em, Request $request): Response
    {
        $form = $this->createForm(RegisterFormType::class);
        $form->handleRequest($request);
        if ($form->isSubmitted() && $form->isValid()) {
            die("Submitted data is valid");
        }
        return $this->render('user/register.html.twig', [
            'registerForm' => $form->createView()
        ]);
    }
}

```

My problem is that, when I submit the form, **no matter what I enter into the Email field** (e.g. "xlsjdjsblsjd" with no "@" in it), `$form->isValid()` always returns true. Why?

I tried changing the type for the `name` field to `IntegerType` and submitting a string, and I do get the expected validation error for that field in that case, so it's not like the whole form is not getting validated at all. It seems like the `EmailType` type considers anything to be valid.

What am I missing?

**Comment by welcoMattic** [comment, marked as answer]
Hi @php4fan, I think you are confusing 2 things which are separated in Symfony : 
FormTypes and Constraints.

Actually `EmailType` has no default Constraint for email validation. If you want that, you have to add yourself an [`Email` constraint](https://symfony.com/doc/current/reference/constraints/Email.html) on your `EmailType`.

**Comment by php4fan** [comment]
I see. But on the other hand, an `IntegerType` type does have some constraint by default that requires it to be an integer (or at the very least a number, I didn't check whether it rejects a non-integer float). Is that right?

Why? What is the good reason behind that distinction?

**Reply by stof** [reply]
Well, the IntegerType is expected to return a PHP integer. So it has to perform some transformation and will report errors when its casting fails (this cannot be left to the validator, as the validator sees the final transformed value, which cannot be generated in such case).
The EmailType does not perform any transformation.
